### DISCUSSION TITLE
Kernel regression

### DISCUSSION THREAD

**Discussion by niklasschmitz** [google-research/dex-lang#539] [Show and tell]
Hi,
I made a small snippet for a 1d kernel ridge regression to get started using array comprehensions, which I found fun and thought others might too; I'm really enjoying dex so far!
```
def regress (kernel: a -> a -> Float) (xs: Nx=>a) (ys: Nx=>Float) : a -> Float =
    gram = for i j. kernel xs.i xs.j + select (i==j) 0.0001 0.0
    alpha = solve gram ys
    predict = \x. sum for i. alpha.i * kernel xs.i x
    predict

def rbf (lengthscale:Float) (x:Float) (y:Float) : Float =
  exp (-0.5 * sq ((x - y) / lengthscale))

predict = regress (rbf 0.2) xs ys
```

Here's the [full gist](https://gist.github.com/niklasschmitz/87e3382b20082e90a5d3f121a0d97443), with the rest of the setup (data,  cg solver) copied from the basis function regression example.

**Comment by apaszke** [comment]
This is awesome! Would you be interested in submitting that as an example to the repo?

**Reply by niklasschmitz** [reply]
Sure ! Opened a PR at #541
